File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label([first, second, third | _]) do
    resistance = (10 * value(first) + value(second)) * 10**value(third)
    if resistance > 1000 do
      {resistance/1000, :kiloohms}
    else
      {resistance, :ohms}
    end
  end

  defp value(color) do
    case color do
      :black -> 0
      :brown -> 1
      :red -> 2
      :orange -> 3
      :yellow -> 4
      :green -> 5
      :blue -> 6
      :violet -> 7
      :grey -> 8
      :white -> 9
    end
  end
end

File: solutions/javascript/resistor-color-trio/1/resistor-color-trio.js


export class ResistorColorTrio {

  static VALUES = {
    black: 0, brown: 1, red: 2, orange: 3, yellow: 4, green: 5, blue: 6, violet: 7,
    grey: 8, white: 9}

  static UNITS = ['ohms', 'kiloohms']
  
  constructor(bands) {

    const valueOf = function(color) {
      const value = ResistorColorTrio.VALUES[color]
      if (value === undefined) throw new Error('invalid color')
      return value
    }
    this._value = (10 * valueOf(bands[0]) + valueOf(bands[1])) * 10**valueOf(bands[2])
  }

  get label() {
    
    let value = this._value
    let unitIndex = 0
    while (value >= 1000) {
      value /= 1000
      unitIndex++
    }
    
    return `Resistor value: ${value} ${ResistorColorTrio.UNITS[unitIndex]}`
  }

}
